src/_pytest/subtests.py

```diff
@@ -0,0 +1,431 @@
+            self.ihook.pytest_exception_interact(
+                node=self.request.node, call=call_info, report=sub_report
+            )
+
+        if exc_val is not None:
+            if self.request.session.shouldfail:
+                return False
+        return True
+
+
+@contextmanager
+def capturing_output(request: SubRequest) -> Iterator[Captured]:
+    option = request.config.getoption("capture", None)
+
+    # capsys or capfd are active, subtest should not capture.
+    capman = request.config.pluginmanager.getplugin("capturemanager")
+    capture_fixture_active = getattr(capman, "_capture_fixture", None)
+
+    if option == "sys" and not capture_fixture_active:
+        fixture = CaptureFixture(SysCapture, request, _ispytest=True)
+    elif option == "fd" and not capture_fixture_active:
+        fixture = CaptureFixture(FDCapture, request, _ispytest=True)
+    else:
+        fixture = None
+
+    if fixture is not None:
+        fixture._start()
+
+    captured = Captured()
+    try:
+        yield captured
+    finally:
+        if fixture is not None:
+            out, err = fixture.readouterr()
+            fixture.close()
+            captured.out = out
+            captured.err = err
+
+
+@contextmanager
+def capturing_logs(
+    request: SubRequest,
+) -> Iterator[CapturedLogs | NullCapturedLogs]:
+    logging_plugin = request.config.pluginmanager.getplugin("logging-plugin")
+    if logging_plugin is None:
+        yield NullCapturedLogs()
+    else:
+        handler = LogCaptureHandler()
+        handler.setFormatter(logging_plugin.formatter)
+
+        captured_logs = CapturedLogs(handler)
+        with catching_logs(handler):
+            yield captured_logs
+
+
+@dataclasses.dataclass
+class Captured:
+    out: str = ""
+    err: str = ""
+
+    def update_report(self, report: TestReport) -> None:
+        if self.out:
+            report.sections.append(("Captured stdout call", self.out))
+        if self.err:
+            report.sections.append(("Captured stderr call", self.err))
+
+
+@dataclasses.dataclass
+class CapturedLogs:
+    handler: LogCaptureHandler
+
+    def update_report(self, report: TestReport) -> None:
+        captured_log = self.handler.stream.getvalue()
+        if captured_log:
+            report.sections.append(("Captured log call", captured_log))
+
+
+class NullCapturedLogs:
+    def update_report(self, report: TestReport) -> None:
+        pass
+
+
+def pytest_report_to_serializable(report: TestReport) -> dict[str, Any] | None:
+    if isinstance(report, SubtestReport):
+        return report._to_json()
+    return None
+
+
+def pytest_report_from_serializable(data: dict[str, Any]) -> SubtestReport | None:
+    if data.get("_report_type") == "SubTestReport":
+        return SubtestReport._from_json(data)
+    return None
+
+
+def pytest_configure(config: Config) -> None:
+    config.pluginmanager.register(SubtestsReporterPlugin(), SubtestsReporterPlugin.NAME)
```

Comment:
Why is this a separate plugin? It seems like it can be on the top-level plugin, using `config.stash` for `contains_failed_subtests`.